### PR TITLE
Refactor direct window list access

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/openrct2/interface/window.c [hunk 1]
```diff
@@ -856,6 +856,22 @@ void window_close_all_except_class(rct_windowclass cls)
...
 }
 
+void window_close_all_except_flags(uint16 flags)
+{
+    for (rct_window * w = g_window_list; w < RCT2_NEW_WINDOW;)
+    {
+        if (!(w->flags & flags))
+        {
+            window_close(w);
+            w = g_window_list;
```
Comment (janisozaur) [inline src/openrct2/interface/window.c]: `window_close` moves all the items following `w` and requires you to start over… if you looped over the windows in reverse order, you wouldn't have to restart and could reduce memory access.

File: src/openrct2/interface/window.c [hunk 2]
```diff
@@ -2582,3 +2598,86 @@ bool window_is_visible(rct_window* w)
...
+
+rct_viewport * window_get_previous_viewport(rct_viewport * current)
+{
+    bool foundPrevious = (current == NULL);
+    rct_window * window = gWindowNextSlot;
+
+    while (true)
+    {
+        window--;
+        if (window < g_window_list)
```
Comment (janisozaur) [inline src/openrct2/interface/window.c]: In other parts of this change you check for `gWindowNextSlot == NULL`, and indeed `gWindowNextSlot` is initialised to `NULL`. If you happen to initialise `window` with `gWindowNextSlot == NULL`, then `window--` in line above will make it point at some high, but invalid offset in memory.

You should add a check for `window != NULL` as a precondition to entering this loop.

File: src/openrct2/interface/window.c [hunk 3]
```diff
@@ -856,6 +856,22 @@ void window_close_all_except_class(rct_windowclass cls)
     }
 }
 
+void window_close_all_except_flags(uint16 flags)
+{
+    for (rct_window * w = g_window_list; w < RCT2_NEW_WINDOW;)
+    {
+        if (!(w->flags & flags))
```
Comment (janisozaur) [inline src/openrct2/interface/window.c]: If you pass more than one flag as argument to this function, having any of them set will trigger this condition. It looks like this is meant behaviour, but it's not clear given the interface, so please add a comment explaining this.

File: src/openrct2/interface/window.c (hunk 1 above)
Comment (marijnvdwerf) [reply src/openrct2/interface/window.c]: So like this?
```c
    for (rct_window * w = RCT2_NEW_WINDOW - 1; w >= g_window_list; w--)
    {
        if (!(w->flags & flags))
        {
            window_close(w);
        }
    }
```

Comment (janisozaur) [reply src/openrct2/interface/window.c]: yes, although there is `RCT2_LAST_WINDOW` already.

Comment (janisozaur) [reply src/openrct2/interface/window.c]: Oh, and while at it, please also change `windows_close_all` to this, I have just run into this bug that I mentioned above :P

**Comment by marijnvdwerf** [comment]
@janisozaur Issues resolved?

**Comment by AaronVanGeffen** [comment]
The changes look good to me. I think you've addressed @janisozaur's concerns nicely.  Everything seems to work as expected, too, so 👍 from me.

**Pull request merged**